### PR TITLE
Feat/add repos

### PR DESCRIPTION
**Pull request by Leeyaoi**

### CODE REVIEW

File: InnoViber.DAL/Interfaces/IChatRepository.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿using InnoViber.DAL.Models;
+
+namespace InnoViber.DAL.Interfaces;
+
+public interface IChatRepository : IGenericRepository<Chat>
+{ public Task<Chat?> GetById(Guid id);
+}
```
Comment (moltenessence) [inline InnoViber.DAL/Interfaces/IChatRepository.cs]: formatting

File: InnoViber.DAL/Interfaces/IGenericRepository.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace InnoViber.DAL.Interfaces;
+
+public interface IGenericRepository<TEntity>
+{
+    public Task<List<TEntity>> GetAll();
```
Comment (moltenessence) [inline InnoViber.DAL/Interfaces/IGenericRepository.cs]: cancellationToken?

File: InnoViber.DAL/Interfaces/IMessageRepository.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿using InnoViber.DAL.Models;
+
+namespace InnoViber.DAL.Interfaces;
+
+internal interface IMessageRepository : IGenericRepository<Message>
+{
+    public Task<Message?> GetById(Guid id);
```
Comment (moltenessence) [inline InnoViber.DAL/Interfaces/IMessageRepository.cs]: cancellationToken?

File: InnoViber.DAL/Interfaces/IUserRepository.cs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+﻿using InnoViber.DAL.Models;
+
+namespace InnoViber.DAL.Interfaces;
+
+public interface IUserRepository : IGenericRepository<User>
+{
+    public Task<User?> GetById(Guid id);
```
Comment (moltenessence) [inline InnoViber.DAL/Interfaces/IUserRepository.cs]: cancellationToken?

File: InnoViber.DAL/Interfaces/IMessageRepository.cs (hunk 3 above)
Comment (moltenessence) [reply InnoViber.DAL/Interfaces/IMessageRepository.cs]: why not to move GetById to generic repo?

File: InnoViber.DAL/Repositories/GenericRepository.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+    public GenericRepository(ViberContext context)
+    {
+        _viberContext = context;
+        _dbSet = context.Set<TEntity>();
+    }
+
+    public Task<List<TEntity>> GetAll()
+    {
+        throw new NotImplementedException();
```
Comment (moltenessence) [inline InnoViber.DAL/Repositories/GenericRepository.cs]: ?

**Review by moltenessence — CHANGES_REQUESTED** [review]

File: InnoViber.DAL/Interfaces/IGenericRepository.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+public interface IGenericRepository<TEntity>
+{
+    public Task<List<TEntity>> GetAll(CancellationToken ct);
+
+    public Task Create(TEntity entity, CancellationToken ct);
+
+    public Task<TEntity> Update(TEntity entity, CancellationToken ct);
+
+    public Task<TEntity?> GetByPredicate(Expression<Func<TEntity, bool>> predicate);
```
Comment (pavelGagarin0870) [inline InnoViber.DAL/Interfaces/IGenericRepository.cs]: Cancelation token

**Review by pavelGagarin0870 — APPROVED** [review]

File: InnoViber.DAL/Interfaces/IGenericRepository.cs (hunk 6 above)
Comment (moltenessence) [inline InnoViber.DAL/Interfaces/IGenericRepository.cs]: cancellation token

File: InnoViber.DAL/Interfaces/IMessageRepository.cs [hunk 7]
```diff
@@ -0,0 +1,7 @@
+﻿using InnoViber.DAL.Models;
+
+namespace InnoViber.DAL.Interfaces;
+
+internal interface IMessageRepository : IGenericRepository<Message>
```
Comment (moltenessence) [inline InnoViber.DAL/Interfaces/IMessageRepository.cs]: why internal?

File: InnoViber.DAL/Repositories/ChatRepository.cs [hunk 8]
```diff
@@ -0,0 +1,17 @@
+﻿using InnoViber.DAL.Data;
+using InnoViber.DAL.Interfaces;
+using InnoViber.DAL.Models;
+using Microsoft.EntityFrameworkCore;
+
+namespace InnoViber.DAL.Repositories;
+
+internal class ChatRepository : GenericRepository<Chat>, IChatRepository
```
Comment (moltenessence) [inline InnoViber.DAL/Repositories/ChatRepository.cs]: why internal?

File: InnoViber.DAL/Repositories/ChatRepository.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+{
+    public ChatRepository(ViberContext context) : base(context)
+    { }
+
+    new public Task<Chat?> GetById(Guid Id) => _dbSet.Where(x => x.Id == Id)
```
Comment (moltenessence) [inline InnoViber.DAL/Repositories/ChatRepository.cs]: why 'new' instead of 'override'?

File: InnoViber.DAL/Repositories/MessageRepository.cs [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+using Microsoft.EntityFrameworkCore;
+
+namespace InnoViber.DAL.Repositories;
+
+internal class MessageRepository : GenericRepository<Message>, IMessageRepository
+{
+    public MessageRepository(ViberContext context) : base(context)
+    { }
+
+    new public Task<Message?> GetById(Guid Id) => _dbSet.Where(x => x.Id == Id)
```
Comment (moltenessence) [inline InnoViber.DAL/Repositories/MessageRepository.cs]: cancellation token

File: InnoViber.DAL/Repositories/UserRepository.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+using Microsoft.EntityFrameworkCore;
+
+namespace InnoViber.DAL.Repositories;
+
+internal class UserRepository : GenericRepository<User>, IUserRepository
+{
+    public UserRepository(ViberContext context) : base(context)
+    {}
+
+    new public Task<User?> GetById(Guid Id) => _dbSet.Where(x => x.Id == Id).Include(x => x.Chats).FirstOrDefaultAsync();
```
Comment (moltenessence) [inline InnoViber.DAL/Repositories/UserRepository.cs]: cancellation token

File: InnoViber.DAL/Repositories/UserRepository.cs [hunk 12]
```diff
@@ -0,0 +1,14 @@
+﻿using InnoViber.DAL.Data;
+using InnoViber.DAL.Interfaces;
+using InnoViber.DAL.Models;
+using Microsoft.EntityFrameworkCore;
+
+namespace InnoViber.DAL.Repositories;
+
+internal class UserRepository : GenericRepository<User>, IUserRepository
```
Comment (moltenessence) [inline InnoViber.DAL/Repositories/UserRepository.cs]: internal

File: InnoViber.DAL/Repositories/MessageRepository.cs [hunk 13]
```diff
@@ -0,0 +1,17 @@
+﻿using InnoViber.DAL.Data;
+using InnoViber.DAL.Interfaces;
+using InnoViber.DAL.Models;
+using Microsoft.EntityFrameworkCore;
+
+namespace InnoViber.DAL.Repositories;
+
+internal class MessageRepository : GenericRepository<Message>, IMessageRepository
```
Comment (moltenessence) [inline InnoViber.DAL/Repositories/MessageRepository.cs]: internal

File: InnoViber.DAL/Repositories/GenericRepository.cs [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+    public Task Delete(TEntity entity, CancellationToken ct)
+    {
+        _dbSet.Remove(entity);
+        return _viberContext.SaveChangesAsync(ct);
+    }
+
+    public Task<TEntity?> GetByPredicate(Expression<Func<TEntity, bool>> predicate)
+    {
+        return _dbSet.Where(predicate).FirstOrDefaultAsync();
```
Comment (moltenessence) [inline InnoViber.DAL/Repositories/GenericRepository.cs]: ct

**Review by moltenessence — CHANGES_REQUESTED** [review]

File: InnoViber.DAL/Repositories/ChatRepository.cs [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+using Microsoft.EntityFrameworkCore;
+
+namespace InnoViber.DAL.Repositories;
+
+public class ChatRepository : GenericRepository<Chat>, IChatRepository
+{
+    public ChatRepository(ViberContext context) : base(context)
+    { }
+
+    override public Task<Chat?> GetById(Guid Id, CancellationToken ct) => _dbSet.Where(x => x.Id == Id)
```
Comment (moltenessence) [inline InnoViber.DAL/Repositories/ChatRepository.cs]: why not public method?

File: InnoViber.DAL/Repositories/UserRepository.cs [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+using Microsoft.EntityFrameworkCore;
+
+namespace InnoViber.DAL.Repositories;
+
+public class UserRepository : GenericRepository<User>, IUserRepository
+{
+    public UserRepository(ViberContext context) : base(context)
+    {}
+
+    override public Task<User?> GetById(Guid Id, CancellationToken ct) => _dbSet.Where(x => x.Id == Id)
```
Comment (moltenessence) [inline InnoViber.DAL/Repositories/UserRepository.cs]: why not public method?

File: InnoViber.DAL/Repositories/MessageRepository.cs [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+using Microsoft.EntityFrameworkCore;
+
+namespace InnoViber.DAL.Repositories;
+
+public class MessageRepository : GenericRepository<Message>, IMessageRepository
+{
+    public MessageRepository(ViberContext context) : base(context)
+    { }
+
+    override public Task<Message?> GetById(Guid Id, CancellationToken ct) => _dbSet.Where(x => x.Id == Id)
```
Comment (moltenessence) [inline InnoViber.DAL/Repositories/MessageRepository.cs]: why not public method?

File: InnoViber.DAL/Interfaces/IGenericRepository.cs [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+
+    public Task Create(TEntity entity, CancellationToken ct);
+
+    public Task<TEntity> Update(TEntity entity, CancellationToken ct);
+
+    public Task<TEntity?> GetByPredicate(Expression<Func<TEntity, bool>> predicate, CancellationToken ct);
+
+    public Task<TEntity?> GetById(Guid Id, CancellationToken ct);
+
+    public Task Delete(TEntity entity, CancellationToken ct);
```
Comment (moltenessence) [inline InnoViber.DAL/Interfaces/IGenericRepository.cs]: you don't need access modifiers for methods in interface

**Review by moltenessence — CHANGES_REQUESTED** [review]

File: InnoViber.DAL/Entities/Message.cs [hunk 19]
```diff
@@ -0,0 +1,12 @@
+﻿namespace InnoViber.DAL.Models;
+
+public class Message : BaseEntity
+{
+    public DateTime Date { get; set; }
+    public string Text { get; set; } = string.Empty;
+    public int Status { get; set; }
```
Comment (moltenessence) [inline InnoViber.DAL/Entities/Message.cs:7]: make it as enum in future

**Review by moltenessence — APPROVED** [review]

**Pull request merged**